### PR TITLE
Airflow kuttl

### PR DESCRIPTION
**Pull request by adwk67**
## Review Checklist
- [ ] Code contains useful comments
- [ ] Changelog updated (or not applicable)

### CODE REVIEW

**Comment by adwk67** [comment]
N.B. draft, so that the CI tests to be changed

**Review by maltesander — COMMENTED** [review]
Tested and finishes successfully. 
`airflow-scheduler-default-0` had an error though? Wasnt fast enough to get the logs. I can try again.

**Comment by maltesander** [comment]
It is not coming up properly though:
```
The above exception was the direct cause of the following exception:                                                   │
│                                                                                                                        │
│ Traceback (most recent call last):                                                                                     │
│   File "/home/airflow/.local/bin/airflow", line 8, in <module>                                                         │
│     sys.exit(main())                                                                                                   │
│   File "/home/airflow/.local/lib/python3.8/site-packages/airflow/__main__.py", line 48, in main                        │
│     args.func(args)                                                                                                    │
│   File "/home/airflow/.local/lib/python3.8/site-packages/airflow/cli/cli_parser.py", line 48, in command               │
│     return func(*args, **kwargs)                                                                                       │
│   File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/cli.py", line 92, in wrapper                    │
│     return f(*args, **kwargs)                                                                                          │
│   File "/home/airflow/.local/lib/python3.8/site-packages/airflow/cli/commands/scheduler_command.py", line 75, in sched │
│     _run_scheduler_job(args=args)                                                                                      │
│   File "/home/airflow/.local/lib/python3.8/site-packages/airflow/cli/commands/scheduler_command.py", line 46, in _run_ │
│     job.run()                                                                                                          │
│   File "/home/airflow/.local/lib/python3.8/site-packages/airflow/jobs/base_job.py", line 241, in run                   │
│     session.commit()                                                                                                   │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 1046, in commit              │
│     self.transaction.commit()                                                                                          │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 504, in commit               │
│     self._prepare_impl()                                                                                               │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 483, in _prepare_impl        │
│     self.session.flush()                                                                                               │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 2540, in flush               │
│     self._flush(objects)                                                                                               │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 2682, in _flush              │
│     transaction.rollback(_capture_exception=True)                                                                      │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__         │
│     compat.raise_(                                                                                                     │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 182, in raise_               │
│     raise exception                                                                                                    │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 2642, in _flush              │
│     flush_context.execute()                                                                                            │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py", line 422, in execute           │
│     rec.execute(self)                                                                                                  │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py", line 586, in execute           │
│     persistence.save_obj(                                                                                              │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py", line 239, in save_obj         │
│     _emit_insert_statements(                                                                                           │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py", line 1135, in _emit_insert_st │
│     result = cached_connections[connection].execute(                                                                   │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1011, in execute             │
│     return meth(self, multiparams, params)                                                                             │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/sql/elements.py", line 298, in _execute_on_connect │
│     return connection._execute_clauseelement(self, multiparams, params)                                                │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1124, in _execute_clauseelem │
│     ret = self._execute_context(                                                                                       │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1316, in _execute_context    │
│     self._handle_dbapi_exception(                                                                                      │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1510, in _handle_dbapi_excep │
│     util.raise_(                                                                                                       │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 182, in raise_               │
│     raise exception                                                                                                    │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context    │
│     self.dialect.do_execute(                                                                                           │
│   File "/home/airflow/.local/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 608, in do_execute        │
│     cursor.execute(statement, parameters)                                                                              │
│ sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedTable) relation "job" does not exist                        │
│ LINE 1: INSERT INTO job (dag_id, state, job_type, start_date, end_da...                                                │
│                     ^                                                                                                  │
│                                                                                                                        │
│ [SQL: INSERT INTO job (dag_id, state, job_type, start_date, end_date, latest_heartbeat, executor_class, hostname, unix │
│ [parameters: {'dag_id': None, 'state': <TaskInstanceState.RUNNING: 'running'>, 'job_type': 'SchedulerJob', 'start_date │
│ (Background on this error at: http://sqlalche.me/e/13/f405)    
```
Webserver and Workers are good. It got restarted 6 times already:
```
kuttl-test-noted-mollusk    airflow-postgresql-0                            ●    1/1              0 Running     10.244.2.13     kind-worker3           8m27s      │
│ kuttl-test-noted-mollusk    airflow-redis-master-0                          ●    1/1              0 Running     10.244.3.13     kind-worker            8m7s       │
│ kuttl-test-noted-mollusk    airflow-redis-replicas-0                        ●    1/1              0 Running     10.244.2.15     kind-worker3           8m7s       │
│ kuttl-test-noted-mollusk    airflow-redis-replicas-1                        ●    1/1              0 Running     10.244.3.15     kind-worker            7m29s      │
│ kuttl-test-noted-mollusk    airflow-redis-replicas-2                        ●    1/1              0 Running     10.244.1.14     kind-worker2           7m1s       │
│ kuttl-test-noted-mollusk    airflow-scheduler-default-0                     ●    0/1              6 Error       10.244.2.17     kind-worker3           6m31s      │
│ kuttl-test-noted-mollusk    airflow-webserver-default-0                     ●    1/1              0 Running     10.244.2.19     kind-worker3           6m31s      │
│ kuttl-test-noted-mollusk    airflow-worker-default-0                        ●    1/1              0 Running     10.244.2.18     kind-worker3           6m31s      │
│ kuttl-test-noted-mollusk    airflow-worker-default-1                        ●    1/1              0 Running     10.244.2.16                               
```

File: airflow-operator/tests/kuttl/smoke/health.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import requests
+import sys
+
+if __name__ == "__main__":
+    result = 0
+    log_level = 'DEBUG'
+    logging.basicConfig(level=log_level, format='%(asctime)s %(levelname)s: %(message)s', stream=sys.stdout)
+
+    http_code = requests.get("http://airflow-webserver-default:8080/api/v1/health").status_code
+    print(f"Test returned code: [{http_code}]")
```
Comment (razvan) [inline airflow-operator/tests/kuttl/smoke/health.py:12]: The HTTP return code is logged by `requests` when using the `DEBUG` level.

**Review by razvan — APPROVED** [review]
lgtm

**Comment by adwk67** [comment]
Will re-work the jobs so that the db-init definitely finds the clusterRef.

**Comment by backstreetkiwi** [comment]
first test in Jenkins worked fine: https://ci.stackable.tech/view/02%20Component%20Tests%20(flex)/job/Airflow%20Operator%20Integration%20Tests%20(flex)/9/

**Comment by adwk67** [comment]
Tested with the operator having fixed the namespace-bug: this time the scheduler does come up:

```
08:22 $ kubectl logs -f airflow-scheduler-default-0 -c airflow --namespace kuttl-test-settled-pegasus
  ____________       _____________
 ____    |__( )_________  __/__  /________      __
____  /| |_  /__  ___/_  /_ __  /_  __ \_ | /| / /
___  ___ |  / _  /   _  __/ _  / / /_/ /_ |/ |/ /
 _/_/  |_/_/  /_/    /_/    /_/  \____/____/|__/
[2022-02-09 07:22:47,717] {scheduler_job.py:596} INFO - Starting the scheduler
[2022-02-09 07:22:47,718] {scheduler_job.py:601} INFO - Processing each file at most -1 times
[2022-02-09 07:22:47,888] {manager.py:163} INFO - Launched DagFileProcessorManager with pid: 8
[2022-02-09 07:22:47,889] {scheduler_job.py:1114} INFO - Resetting orphaned tasks for active dag runs
[2022-02-09 07:22:47,891] {settings.py:52} INFO - Configured default timezone Timezone('UTC')
[2022-02-09 07:22:56,202] {scheduler_job.py:161} INFO - Exiting gracefully upon receiving signal 15
[2022-02-09 07:22:57,207] {process_utils.py:120} INFO - Sending Signals.SIGTERM to group 8. PIDs of all processes in the group: [8]
[2022-02-09 07:22:57,207] {process_utils.py:75} INFO - Sending the signal Signals.SIGTERM to group 8
[2022-02-09 07:22:57,344] {process_utils.py:70} INFO - Process psutil.Process(pid=8, status='terminated', exitcode=0, started='07:22:47') (8) terminated with exit code 0
[2022-02-09 07:22:57,344] {process_utils.py:120} INFO - Sending Signals.SIGTERM to group 8. PIDs of all processes in the group: []
[2022-02-09 07:22:57,344] {process_utils.py:75} INFO - Sending the signal Signals.SIGTERM to group 8
[2022-02-09 07:22:57,344] {process_utils.py:89} INFO - Sending the signal Signals.SIGTERM to process 8 as process group is missing.
[2022-02-09 07:22:57,345] {scheduler_job.py:655} INFO - Exited execute loop
(base) ✔ ~/gitrepos/integration-tests/airflow-operator [airflow-kuttl ↑·9|✔] 
```

**Comment by backstreetkiwi** [comment]
Reminds me of this: https://i.pinimg.com/1200x/17/b0/5c/17b05c4a80372379ab518832a21fa83c.jpg

> Tested with the operator having fixed the namespace-bug: this time the scheduler does come up:
> 
> ```
> 08:22 $ kubectl logs -f airflow-scheduler-default-0 -c airflow --namespace kuttl-test-settled-pegasus
>   ____________       _____________
>  ____    |__( )_________  __/__  /________      __
> ____  /| |_  /__  ___/_  /_ __  /_  __ \_ | /| / /
> ___  ___ |  / _  /   _  __/ _  / / /_/ /_ |/ |/ /
>  _/_/  |_/_/  /_/    /_/    /_/  \____/____/|__/
> [2022-02-09 07:22:47,717] {scheduler_job.py:596} INFO - Starting the scheduler
> [2022-02-09 07:22:47,718] {scheduler_job.py:601} INFO - Processing each file at most -1 times
> [2022-02-09 07:22:47,888] {manager.py:163} INFO - Launched DagFileProcessorManager with pid: 8
> [2022-02-09 07:22:47,889] {scheduler_job.py:1114} INFO - Resetting orphaned tasks for active dag runs
> [2022-02-09 07:22:47,891] {settings.py:52} INFO - Configured default timezone Timezone('UTC')
> [2022-02-09 07:22:56,202] {scheduler_job.py:161} INFO - Exiting gracefully upon receiving signal 15
> [2022-02-09 07:22:57,207] {process_utils.py:120} INFO - Sending Signals.SIGTERM to group 8. PIDs of all processes in the group: [8]
> [2022-02-09 07:22:57,207] {process_utils.py:75} INFO - Sending the signal Signals.SIGTERM to group 8
> [2022-02-09 07:22:57,344] {process_utils.py:70} INFO - Process psutil.Process(pid=8, status='terminated', exitcode=0, started='07:22:47') (8) terminated with exit code 0
> [2022-02-09 07:22:57,344] {process_utils.py:120} INFO - Sending Signals.SIGTERM to group 8. PIDs of all processes in the group: []
> [2022-02-09 07:22:57,344] {process_utils.py:75} INFO - Sending the signal Signals.SIGTERM to group 8
> [2022-02-09 07:22:57,344] {process_utils.py:89} INFO - Sending the signal Signals.SIGTERM to process 8 as process group is missing.
> [2022-02-09 07:22:57,345] {scheduler_job.py:655} INFO - Exited execute loop
> (base) ✔ ~/gitrepos/integration-tests/airflow-operator [airflow-kuttl ↑·9|✔] 
> ```

File: airflow-operator/tests/kuttl/smoke/health.py (hunk 1 above)
Comment (adwk67) [reply airflow-operator/tests/kuttl/smoke/health.py]: The print line has been removed in the latest commit.

**Review by maltesander — COMMENTED** [review]
First test failed (time out);     case.go:363: resource StatefulSet:kuttl-test-large-mongrel/airflow-redis-replicas: .status.readyReplicas: value mismatch, expected: 3 != actual: 2

Timeout is on 240s, maybe we could increase that?

Second try (with packages downloaded) successful.

**Review by maltesander — APPROVED** [review]
Worked first try! LGTM.

**Comment by backstreetkiwi** [comment]
(oops, already merged)
Test successful in Jenkins: https://ci.stackable.tech/job/Airflow%20Operator%20Integration%20Tests%20(flex)/10/
@adwk67 One thing I am wondering about: PostgreSQL and Redis are deployed by the kuttl scripts. In the script wrapping the test for Jenkins, I am doing it as well. It seems to be working, but it is redundant. I'd be happy to remove it from the surrounding scripts, the more self-contained the tests are, the better

**Comment by adwk67** [comment]
@backstreetkiwi yes, that's a good idea: I left it as it was to test that namespace resolution works properly (i.e. that the secret is used against the correct postgres instance), but it's probably best to remove anything that is redundant.

**Pull request merged**